File: apps/dot/backend/app/routers/telemetry.py
"""Telemetria: eventos anonimos del cliente (sin datos personales)."""
from __future__ import annotations

import logging

from fastapi import APIRouter, Request, Response
from pydantic import BaseModel

log = logging.getLogger("dot.telemetry")

router = APIRouter(prefix="/v1/telemetry", tags=["telemetry"])


class TelemetryEvent(BaseModel):
    type: str
    timestamp: str
    meta: dict = {}


@router.post("/event", status_code=204)
def telemetry_event(request: Request, body: TelemetryEvent):
    """Recibe evento de telemetria anonimo del cliente."""
    if body.type not in ("session_error", "api_latency", "provider_failure", "login_failure"):
        log.warning("Telemetry event type desconocido: %s", body.type)
        return Response(status_code=204)

    log.debug(
        "Telemetry [%s] meta=%s ip=%s",
        body.type,
        {k: v for k, v in body.meta.items() if k != "detail" or len(str(v)) < 100},
        request.client.host if request.client else "?",
    )

    return Response(status_code=204)

File: apps/dot/frontend/src/lib/telemetry.ts
/**
 * Telemetria minima del lado cliente.
 * Solo errores de sesion, latencia de API, fallos de proveedor.
 * Sin datos personales.
 */

import { getApiBaseUrl } from '@/lib/api/base-url'

const TELEMETRY_PATH = '/v1/telemetry/event'

/** URL absoluta del endpoint de telemetría (usa la misma base que apiFetchJson). */
export function resolveTelemetryEventUrl(): string {
  const base = getApiBaseUrl().replace(/\/$/, '')
  return `${base}${TELEMETRY_PATH}`
}

type TelemetryEvent = {
  type: 'session_error' | 'api_latency' | 'provider_failure' | 'login_failure'
  timestamp: string
  meta: Record<string, string | number | boolean | null>
}

let enabled = true

export function disableTelemetry() {
  enabled = false
}

export function enableTelemetry() {
  enabled = true
}

function send(event: TelemetryEvent) {
  if (!enabled) return
  try {
    const url = resolveTelemetryEventUrl()
    const payload = JSON.stringify(event)
    if (navigator.sendBeacon) {
      const blob = new Blob([payload], { type: 'application/json' })
      navigator.sendBeacon(url, blob)
    } else {
      fetch(url, {
        method: 'POST',
        body: payload,
        headers: { 'Content-Type': 'application/json' },
        keepalive: true,
      }).catch(() => {
        /* fallo silencioso */
      })
    }
  } catch {
    /* ignorar errores de telemetria */
  }
}

export function trackSessionError(errorType: string, detail?: string) {
  send({
    type: 'session_error',
    timestamp: new Date().toISOString(),
    meta: { errorType: errorType, detail: detail ?? '' },
  })
}

export function trackApiLatency(endpoint: string, durationMs: number, status: number) {
  send({
    type: 'api_latency',
    timestamp: new Date().toISOString(),
    meta: { endpoint, durationMs: Math.round(durationMs), status },
  })
}

export function trackProviderFailure(providerId: string, errorMessage: string) {
  send({
    type: 'provider_failure',
    timestamp: new Date().toISOString(),
    meta: { providerId, errorMessage: errorMessage.slice(0, 200) },
  })
}

export function trackLoginFailure(reason: string) {
  send({
    type: 'login_failure',
    timestamp: new Date().toISOString(),
    meta: { reason },
  })
}
